oas3.yaml

```diff
@@ -0,0 +1,328 @@
+      operationId: pet_v1
+      parameters:
+        - name: id
+          in: path
+          description: The unique ID of the idea you wish to retrieve.
+          required: true
+          schema:
+            type: string
+            pattern: ^[a-f0-9]{32}$
+      responses:
+        200:
+          description: The pet which you requested information about.
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/PetV1"
+        404:
+          description: The pet could not be found.
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/Error"
+        500:
+          description: An error occurred on the server.
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/Error"
+
+    patch:
+      tags:
+        - pets
+      summary: Mofify Pet (v1)
+      description: Modifies the information about a pet
+      operationId: pet_modify_v1
+      parameters:
+        - name: id
+          in: path
+          description: The unique ID of the idea you wish to retrieve.
+          required: true
+          schema:
+            type: string
+            pattern: ^[a-f0-9]{32}$
+      requestBody:
+        description: The changes you wish to make to the pet's information.
+        content:
+          application/json:
+            schema:
+              $ref: "#/components/schemas/PetV1"
+            example:
+              status: adopted
+
+      responses:
+        200:
+          description: The newly modified pet's information.
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/PetV1"
+        404:
+          description: The pet could not be found.
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/Error"
+        500:
+          description: An error occurred on the server.
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/Error"
+
+components:
+  schemas:
+    HealthV1:
+      required:
+        - status
+      type: object
+      properties:
+        status:
+          type: string
+          description: The current aggregate status of the service.
+          enum:
+            - Healthy
+            - Degraded
+            - Unhealthy
+        results:
+          type: object
+          additionalProperties:
+            type: object
+            required:
+              - status
+              - description
+            properties:
+              status:
+                type: string
+                description: The current status of the service's component.
+                enum:
+                  - Healthy
+                  - Degraded
+                  - Unhealthy
+              description:
```

Comment:
If description is meant to explain why the service is unhealthy, also consider including a text based code in the response that matches the description.  E.g. if the description is "The database cannot be reached", the error code may be "DatabaseUnreachable".  Having a fixed list of these, makes it easy for UI consumers to take action based on the code, like retry in some cases, or turning it into a message that is more user friendly than the technical message in the description field.  Again, this would just be more for a design discussion point than an expectation that we'll get round to coding error handling against it.

Reply:
I see you have a type field that is a uri below for the same concept for error responses.  That sounds good to me too.

Reply:
Good idea, in this case the Health API isn't really intended to be consumed by the frontend however. Instead, it is intended to be consumed by whatever orchestration framework you are using to run the service as a means of ensuring that the service is running correctly and (if needed) isolating unhealthy nodes or repairing them automatically. The specific fields here are just a means of describing which upstream dependencies may be contributing to an unhealthy state and could be used to indicate to your orchestrator (or monitoring stack) why you are reporting an unhealthy/degraded state to aid in troubleshooting.

I would not expect the frontend to use this API under normal situations (unless you're planning on building a status page).

Reply:
As for the `Error` type, it is indeed intended to be consumed by the frontend and the `type` field is intended for exactly that purpose (in combination with the `status` code).

Reply:
Sounds good.  In my day job we consume this type of endpoint in our UI client as well where a service may be temporarily unhealthy while a failover is happening and we want to pre-emptively show users that they cannot currently use the site.  So even if they haven't made a service call in a while we would poll something like this and bring down the shutters when it becomes unavailable.  This is a pretty advanced/niche scenario though and not something we'd probably have time to get into, so no worries.  Just adding this info for context.

Reply:
That's actually a really interesting use of it, I've not personally used it in that context before so really good to get that added perspective. Thanks for sharing and maybe we can add it into the 2nd/4th session as a tidbit on how to monitor and handle failure effectively?

Reply:
Sounds good.  Like I said though, it's a bit niche so not that important for the workshop.

Reply:
I would suggest we don't give them health endpoint for the first session but let them build the endpoint for the fourth session if there is space for that